cpp/open3d/pipelines/registration/Registration.cpp

```diff
@@ -194,152 +192,160 @@ RegistrationResult RegistrationRANSACBasedOnCorrespondence(
-                                        0, num_similar_features - 1)];
-                    }
+                    ransac_corres[j] = corres[utility::UniformRandInt(
+                            0, static_cast<int>(corres.size()) - 1)];
                 }
+
+                Eigen::Matrix4d transformation =
+                        estimation.ComputeTransformation(source, target,
+                                                         ransac_corres);
+
+                // Check transformation: inexpensive
                 bool check = true;
                 for (const auto &checker : checkers) {
-                    if (!checker.get().require_pointcloud_alignment_ &&
-                        !checker.get().Check(source, target, ransac_corres,
-                                             transformation)) {
-                        check = false;
-                        break;
-                    }
-                }
-                if (!check) continue;
-                transformation = estimation.ComputeTransformation(
-                        source, target, ransac_corres);
-                check = true;
-                for (const auto &checker : checkers) {
-                    if (checker.get().require_pointcloud_alignment_ &&
-                        !checker.get().Check(source, target, ransac_corres,
+                    if (!checker.get().Check(source, target, ransac_corres,
                                              transformation)) {
                         check = false;
                         break;
                     }
                 }
                 if (!check) continue;
+
                 geometry::PointCloud pcd = source;
                 pcd.Transform(transformation);
-                auto this_result = GetRegistrationResultAndCorrespondences(
-                        pcd, target, kdtree, max_correspondence_distance,
+                auto result = EvaluateRANSACBasedOnCorrespondence(
+                        pcd, target, corres, max_correspondence_distance,
                         transformation);
-                if (this_result.fitness_ > result_private.fitness_ ||
-                    (this_result.fitness_ == result_private.fitness_ &&
-                     this_result.inlier_rmse_ < result_private.inlier_rmse_)) {
-                    result_private = this_result;
-                }
-#pragma omp critical
-                {
-                    total_validation = total_validation + 1;
-                    if (total_validation >= criteria.max_validation_)
-                        finished_validation = true;
+
+                if (result.IsBetterRANSACThan(best_result_local)) {
+                    best_result_local = result;
+
+                    // Update exit condition if necessary
+                    double exit_itr_d =
+                            std::log(1.0 - criteria.confidence_) /
+                            std::log(1.0 - std::pow(result.fitness_, ransac_n));
```

Comment:
One more thing, sorry. This can relatively easy overflow, and fitness can be zero, sometimes. Shouldn't there be some guard around it?